src/hb-buffer.h

```diff
@@ -76,18 +76,67 @@ typedef struct hb_glyph_info_t {
  * @HB_GLYPH_FLAG_UNSAFE_TO_BREAK: Indicates that if input text is broken at the
  * 				   beginning of the cluster this glyph is part of,
  * 				   then both sides need to be re-shaped, as the
- * 				   result might be different.  On the flip side,
- * 				   it means that when this flag is not present,
- * 				   then it's safe to break the glyph-run at the
- * 				   beginning of this cluster, and the two sides
- * 				   represent the exact same result one would get
- * 				   if breaking input text at the beginning of
- * 				   this cluster and shaping the two sides
- * 				   separately.  This can be used to optimize
- * 				   paragraph layout, by avoiding re-shaping
- * 				   of each line after line-breaking, or limiting
- * 				   the reshaping to a small piece around the
- * 				   breaking point only.
+ * 				   result might be different.
+ *
+ * 				   On the flip side, it means that when this
+ * 				   flag is not present, then it is safe to break
+ * 				   the glyph-run at the beginning of this
+ * 				   cluster, and the two sides will represent the
+ * 				   exact same result one would get if breaking
+ * 				   input text at the beginning of this cluster
+ * 				   and shaping the two sides separately.
+ *
+ * 				   This can be used to optimize paragraph
+ * 				   layout, by avoiding re-shaping of each line
+ * 				   after line-breaking.
+ *
+ * @HB_GLYPH_FLAG_UNSAFE_TO_CONCAT: Indicates that if input text is changed on one
+ * 				   side of the beginning of the cluster this glyph
+ * 				   is part of, then the shaping results for the
+ * 				   other side might change.
+ *
+ * 				   Note that the absence of this flag will NOT by
+ * 				   itself mean that it IS safe to concat text.
+ * 				   Only two pieces of text both of which clear of
+ * 				   this flag can be concatenated safely.
+ *
+ * 				   This can be used to optimize paragraph
+ * 				   layout, by avoiding re-shaping of each line
+ * 				   after line-breaking, by limiting the
+ * 				   reshaping to a small piece around the
+ * 				   breaking positin only, even if the breaking
+ * 				   position carries the
+ * 				   #HB_GLYPH_FLAG_UNSAFE_TO_BREAK or when
+ * 				   hyphenation or other text transformation
+ * 				   happens at line-break position, in the following
+ * 				   way:
+ *
+ * 				   1. Iterate back from the line-break position till
+ * 				   the the first cluster start position that is
+ * 				   NOT unsafe-to-concat, 2. shape the segment from
```

Comment:
Can you elaborate a bit more how UNSAFE_TO_BREAK can't do what's explained here, but UNSAFE_TO_CONCAT can? I was under an impression that UNSAFE_TO_BREAK is for line breaking, and UNSAFE_TO_CONCAT is for editing/appending.

Currently, we use UNSAFE_TO_BREAK for the line breaking purpose, and use heuristic rules for when editing/appending. I can understand UNSAFE_TO_CONCAT is useful when editing/appending, including when appending hyphenation while line braking, but in this step 1, I thought UNSAFE_TO_BREAK is the flag to use.

It'd be really helpful if the comment can cover in what cases using UNSAFE_TO_BREAK in step 1 can create a wrong result.

Reply:
> Can you elaborate a bit more how UNSAFE_TO_BREAK can't do what's explained here, but UNSAFE_TO_CONCAT can? I was under an impression that UNSAFE_TO_BREAK is for line breaking, and UNSAFE_TO_CONCAT is for editing/appending.

Hi @kojiishi. Please see https://github.com/harfbuzz/harfbuzz/issues/1463#issuecomment-448081494

Basically, if the point of line-break is UNSAFE_TO_BREAK, then you need this UNSAFE_TO_CONCAT for optimized line-breaking. You cannot use the UNSAFE_TO_BREAK for optimized line-breaking at that point.